### PR TITLE
Update pipeline editor to use relative paths

### LINKED ISSUE #608: "No such file or directory" error when deploying a nonlocally created pipeline
If I check out a pipeline and try to run it on KFP I get the error below since the path

> /Users/romeokienzler/Documents/proj/1codait/corona/analysis/covid-notebooks/covid-notebooks/notebooks

is not valid but

> /Users/romeokienzler/Documents/proj/1codait/corona/analysis/covid-notebooks/notebooks

is. Note that the pipeline file resides in 

> /Users/romeokienzler/Documents/proj/1codait/corona/analysis/covid-notebooks/pipelines

Ways to reproduce
1. git clone https://github.com/CODAIT/covid-notebooks.git
2. cd covid-notebooks
3. jupyter-lab
4. In browser, open pipelines/eu_data.pipeline and try to run

> Traceback (most recent call last):
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 147, in process
>     kfp.compiler.Compiler().compile(cc_pipeline, pipeline_path)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 879, in compile
>     package_path=package_path)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 941, in _create_and_write_workflow
>     pipeline_conf)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 785, in _create_workflow
>     pipeline_func(*args_list)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 121, in cc_pipeline
>     dependency_archive_path = self._generate_dependency_archive(operation)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 196, in _generate_dependency_archive
>     recursive=operation.recursive_dependencies)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/util/archive.py", line 80, in create_temp_archive
>     tar.add(source_dir, arcname="", filter=tar_filter)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/tarfile.py", line 1928, in add
>     tarinfo = self.gettarinfo(name, arcname)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/tarfile.py", line 1807, in gettarinfo
>     statres = os.lstat(name)
> FileNotFoundError: [Errno 2] No such file or directory: '/Users/romeokienzler/Documents/proj/1codait/corona/analysis/covid-notebooks/covid-notebooks/notebooks'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/tornado/web.py", line 1701, in _execute
>     result = method(*self.path_args, **self.path_kwargs)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/scheduler/handler.py", line 39, in post
>     run_url = PipelineProcessorManager.process(pipeline)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/pipeline/processor.py", line 59, in process
>     return processor.process(pipeline)
>   File "/Users/romeokienzler/opt/anaconda3/envs/tf2.0/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 152, in process
>     format(pipeline_name, pipeline_path), str(ex))
> RuntimeError: ('Error compiling pipeline romeotest-0605103914 at /var/folders/95/7hxgmrh92z91hylbxw647jhr0000gn/T/tmp0ejxy3vi/romeotest-0605103914.tar.gz', "[Errno 2] No such file or directory: '/Users/romeokienzler/Documents/proj/1codait/corona/analysis/covid-notebooks/covid-notebooks/notebooks'")

### LINKED ISSUE #659: pipelines store notebook path information that causes issues
To recreate:
- Assume JupyterLab is launched in `/mydir`. `mydir` therefore is the working directory, which is empty. 
- Create a subdirectory in `mydir` named `mydir2` (`/mydir/mydir2`)
- Create a notebook `mynotebook` in `/mydir/mydir2`
- Create a pipeline `mypipeline` in `/mydir/mydir2`
- Add notebook `mynotebook` to the pipeline
- Save the pipeline
- Inspect the saved pipeline file. 

You'll notice that the notebook's `filename` is stored as `mydir2/mynotebook.ipynb`

```
[
        {
          "id": "424905cf-aded-438b-a357-5b9572efc847",
          "type": "execution_node",
          "op": "execute-notebook-node",
          "app_data": {
            "filename": "mydir2/mynotebook.ipynb",
          ...
```

As a result the pipeline is not portable:
- Quit JupyterLab
- Copy the pipeline file and the notebook file (which reside in the same directory) to a different directory, e.g. `/mydir3`.
- Open Jupyterlab
- Open the pipeline. It is now unusable because it references a notebook in the `mydir2` directory that does not exist.

### LINKED ISSUE #823: Update pipeline editor to use relative paths
Subtask of #822 

Currently all file paths used by the pipeline editor are relative to the jupyter lab working dir (notebook_dir). We want to change this so that the file paths for nodes in a pipeline are relative to the pipeline file, similar to how python imports work in notebooks.

As part of this update we will need to also make sure any relative paths are converted to absolute paths prior to submission to the server.

File paths would need to be validated/converted at the following steps:
- When opening
- When drag-and-drop
- When submitting the pipeline

### LINKED ISSUE #825: Update pipeline migration to handle relative paths
Part of #822 and follow up to #823 (may be implemented at the same time)

After the update to relative paths we should increment the pipeline file version and add updating the paths to be relative to the migration functionality.

### PR DESCRIPTION
**Pull request by ajbozarth**
Currently nodes in a pipeline save thier paths relative to the
jupyter lab workspace. This update them to use paths relative to
the pipeline file instead. This enables moving a pipeline "project",
such as moving the directory a pipelines contents are in, checking
out a project from git, and starting lab from a different workspace.

A node's path is converted to be relative to the pipeline upon add
and that relative path is what's stored in the pipeline file. The
path is converted back to be relative to the workspace prior to
being submitted to the export or schedule api, or when a node is
"opened"

Note that this change is designed to work in conjunction with
validation of the node files and the ability to change node file
locations.

Fixes #823
Fixes #825 

Fixes #659 
Fixes #608 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by marthacryan** [comment]
Just tried this with [covid-notebooks](https://github.com/CODAIT/covid-notebooks) (both `eu_data` and `us_data`) and got the following error:
<details>
  <summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 220, in _cc_pipeline
    dependency_archive_path = self._generate_dependency_archive(operation)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 275, in _generate_dependency_archive
    require_complete=True)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/util/archive.py", line 119, in create_temp_archive
    tar.add(source_dir, arcname="", filter=tar_filter)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/tarfile.py", line 1928, in add
    tarinfo = self.gettarinfo(name, arcname)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/tarfile.py", line 1807, in gettarinfo
    statres = os.lstat(name)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/martha.cryanibm.com/gitrepos/elyra/notebooks'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 59, in process
    kfp.compiler.Compiler().compile(pipeline_function, pipeline_path)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 926, in compile
    allow_telemetry=allow_telemetry)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 983, in _create_and_write_workflow
    allow_telemetry)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 806, in _create_workflow
    pipeline_func(*args_list)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 57, in <lambda>
    pipeline_function = lambda: self._cc_pipeline(pipeline, pipeline_name)  # nopep8 E731
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 238, in _cc_pipeline
    format(operation.name, ex))
FileNotFoundError: Node 'etl_eu_data' referenced dependencies that were not found: [Errno 2] No such file or directory: '/Users/martha.cryanibm.com/gitrepos/elyra/notebooks'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/tornado/web.py", line 1703, in _execute
    result = await result
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/handlers.py", line 89, in post
    response = await PipelineProcessorManager.instance().process(pipeline)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor.py", line 70, in process
    res = await asyncio.get_event_loop().run_in_executor(None, processor.process, pipeline)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/martha.cryanibm.com/opt/anaconda3/envs/elyra/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 65, in process
    format(pipeline_name, pipeline_path), str(ex)) from ex
RuntimeError: ('Error compiling pipeline h-0814113152 at /var/folders/qh/9xpk2lt54zxg64zrb_xmthnr0000gn/T/tmpjjz9dcz3/h-0814113152.tar.gz', "Node 'etl_eu_data' referenced dependencies that were not found: [Errno 2] No such file or directory: '/Users/martha.cryanibm.com/gitrepos/elyra/notebooks'")
```
</details>

Thoughts? The pipeline files are in `/Users/martha.cryanibm.com/gitrepos/elyra/covid-notebooks-master/pipelines`. It seems like it should be using the path `/Users/martha.cryanibm.com/gitrepos/elyra/covid-notebooks-master/notebooks` and is instead using the path `/Users/martha.cryanibm.com/gitrepos/elyra/notebooks`.

**Comment by kevin-bates** [comment]
I'm seeing similar issues.  My notebook dir is `/Users/kbates/notebooks/elyra`.  My pipline and its notebooks are all contained in `/pipeline1` (absolute path = `/Users/kbates/notebooks/elyra/pipeline1`), yet I get this when attempting to run the pipeline:

<details>
  <summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 220, in _cc_pipeline
    dependency_archive_path = self._generate_dependency_archive(operation)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 275, in _generate_dependency_archive
    require_complete=True)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/util/archive.py", line 119, in create_temp_archive
    tar.add(source_dir, arcname="", filter=tar_filter)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/tarfile.py", line 1928, in add
    tarinfo = self.gettarinfo(name, arcname)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/tarfile.py", line 1807, in gettarinfo
    statres = os.lstat(name)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/kbates/notebooks/elyra/pipeline1/pipeline1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 59, in process
    kfp.compiler.Compiler().compile(pipeline_function, pipeline_path)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 926, in compile
    allow_telemetry=allow_telemetry)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 983, in _create_and_write_workflow
    allow_telemetry)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/kfp/compiler/compiler.py", line 806, in _create_workflow
    pipeline_func(*args_list)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 57, in <lambda>
    pipeline_function = lambda: self._cc_pipeline(pipeline, pipeline_name)  # nopep8 E731
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 238, in _cc_pipeline
    format(operation.name, ex))
FileNotFoundError: Node 'node1' referenced dependencies that were not found: [Errno 2] No such file or directory: '/Users/kbates/notebooks/elyra/pipeline1/pipeline1'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/tornado/web.py", line 1703, in _execute
    result = await result
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/handlers.py", line 89, in post
    response = await PipelineProcessorManager.instance().process(pipeline)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor.py", line 70, in process
    res = await asyncio.get_event_loop().run_in_executor(None, processor.process, pipeline)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/anaconda3/envs/elyra-dev/lib/python3.7/site-packages/elyra/pipeline/processor_kfp.py", line 65, in process
    format(pipeline_name, pipeline_path), str(ex)) from ex
RuntimeError: ('Error compiling pipeline aj-0814133826 at /var/folders/4w/qxgsbhyx1y93qlk9mnzwjryc0000gn/T/tmpuw6tvuxt/aj-0814133826.tar.gz', "Node 'node1' referenced dependencies that were not found: [Errno 2] No such file or directory: '/Users/kbates/notebooks/elyra/pipeline1/pipeline1'")
```
</details>

**Comment by kevin-bates** [comment]
Creating a fresh pipeline appears to function correctly.  I then took the directory containing that pipeline and relocated it to a different part of the tree and restarted Lab with a different root directory (that includes the new location) and submission worked there as well (nice!).

So I think the above issues are more about handling existing pipelines in which the notebook references are relative to the root directory than the pipeline file.

Opening the pipeline-relative notebook from the pipeline editor also works.  :+1:

**Comment by ajbozarth** [comment]
@marthacryan and @kevin-bates I believe you're hitting an edge case in my implementation that @lresende and I agree was ok to leave. The pipeline converter that updates old pipeline will malformed any pipeline that does not currently work in your environment. So if you checked out the covid pipeline locally and it worked in master then it should work after the coversion in this branch. But if you followed the same steps on master and the pipeline wouldn't work, then the conversion in this branch would malformed the pipeline paths. If that doesn't make sense then I'm willing to hop onto a Webex to explain better.

**Comment by kevin-bates** [comment]
After attempting the same scenario as what failed before, I could not reproduce the issue.  So I suspect I had adjusted my root directory unknowingly when running the pipeline the first time.  Nice stuff.

Since it's now the case that pipeline files must be moved with and exist in the same relative location as their referenced notebooks, we should update [this section](https://elyra.readthedocs.io/en/latest/user_guide/pipelines.html#distributing-your-pipeline) of the docs to reflect how sharing/distribution would now occur.

**Comment by ajbozarth** [comment]
rebased off of latest master, including the required update referenced in https://github.com/elyra-ai/elyra/pull/860#discussion_r470361168

**Comment by kevin-bates** [comment]
> Aside from the edge case

Just to make sure.  The 'edge case' is that you copied the covid-notebooks repo to some location such that it works or doesn't work using master?

Looking at your comment above, I believe it boils down to - which directory did you start Lab (or point `--notebook-dir`)?  If you started/pointed in `/Users/martha.cryanibm.com/gitrepos/elyra/covid-notebooks-master`, I would expect the covid pipeline to run against BOTH master and this branch.  If you started in `/Users/martha.cryanibm.com/gitrepos/elyra`, then I would expect to see the issue you found because the locations of the notebooks in the pipeline file are incorrect (for Master and thus for this branch).

**Comment by marthacryan** [comment]
@kevin-bates It's the latter (so it's expected to see this issue)

**Comment by kevin-bates** [comment]
Ok - thank you for the clarification.

**Comment by ajbozarth** [comment]
I just pushed an update to the doc and also added support for the recently merged #867 

The docs updates cover the changes for both this and the other recently merged PRs under the same umbrella issue.

@karlaspuldaro @marthacryan I did not include updates to the images or gifs in the docs which are now out of date since #867 was merged. Would one of you be able to update those since you've done created them previously? (can be handled in a follow up)

File: docs/source/user_guide/pipelines.md [hunk 1]
```diff
@@ -70,31 +70,15 @@ needed for the pipeline are present.
...
+When deploying a shared or distributed pipeline repository or archive, it is **very important** that the pipeline notebooks be extracted into the **same location relative to the pipeline file**.
 
 ##### Confirming Notebook Locations
-An easy way to confirm that your extracted pipeline's notebook references are correctly located is to double-click the notebook node in your pipeline and ensure the notebook appears in an adjacent tab.  If nothing happens or a `404` error shows on the server's console, the notebook is not properly positioned relative to the server's workspace, and attempts to run or export the pipeline will fail.
-
-At this time, the only recourse, should it not be possible to fix the relative location via redeployment of the pipeline, would be to edit each of the `filename:` fields in the `.pipeline` file such that notebook locations are satisfied, then refresh the browser.
+Pipeline validation checks for the existence of the notebook file associated with each node upon opening the editor and will highlight any nodes with missing files. If a file is missing or in an unexpected location the file location can be changed using the
```
Comment (karlaspuldaro) [inline docs/source/user_guide/pipelines.md:81]: incomplete sentence?

Comment (kevin-bates) [inline docs/source/user_guide/pipelines.md:81]:
```suggestion
Pipeline validation checks for the existence of the notebook file associated with each node upon opening the editor and will highlight any nodes with missing files. If a file is missing or in an unexpected location the file location can be changed using the adjacent `Browse` button.
```

**Comment by ptitzler** [comment]
We should indicate in the change log that after upgrade (to a version that includes this PR) pipelines must be migrated and that pipelines created with Elyra versions that include this PR won't run on older versions.

**Comment by kevin-bates** [comment]
> (Not introduced by this PR) In user_guide/pipelines.md can you fix the incomplete sentence:
>> If a file is missing or in an unexpected location the file location can be changed using the

This was introduced by this PR (as there was the only resource previously was to manually edit the file) and addressed in this commit: https://github.com/elyra-ai/elyra/pull/861/commits/cd4c8cc9a780d58959fd2a2834c978c723924b82

**Pull request merged**